working/0678 - partial classes/feature-brainstorm.md

```diff
@@ -0,0 +1,199 @@
+
+## json_serializable before/after example
+
+Here's a motivating example with a speculative future implementation for
+[json_serializable](https://pub.dev/packages/json_serializable).
+
+### Before (current behavior)
+
+```dart
+// before: User-generated code
+import 'package:json_annotation/json_annotation.dart';
+
+part 'example.g.dart';
+
+@JsonSerializable(nullable: false)
+class Person {
+  final String firstName;
+  final String lastName;
+  final DateTime dateOfBirth;
+  Person({this.firstName, this.lastName, this.dateOfBirth});
+
+  // User needs to connect machine-generated factory. Cannot be a constructor.
+  factory Person.fromJson(Map<String, dynamic> json) => _$PersonFromJson(json);
+
+  // User needs to connect machine-generated function.
+  Map<String, dynamic> toJson() => _$PersonToJson(this);
+}
+```
+
+```dart
+// before: machine-generated
+part of 'example.dart';
+
+// Machine-generated code dirties up the private namespace of the user library
+// with specially named private functions
+Person _$PersonFromJson(Map<String, dynamic> json) {
+  return Person(
+    firstName: json['firstName'] as String,
+    lastName: json['lastName'] as String,
+    dateOfBirth: DateTime.parse(json['dateOfBirth'] as String),
+  );
+}
+
+Map<String, dynamic> _$PersonToJson(Person instance) => <String, dynamic>{
+      'firstName': instance.firstName,
+      'lastName': instance.lastName,
+      'dateOfBirth': instance.dateOfBirth.toIso8601String(),
+    };
+```
+
+### After (possible future)
+
+```dart
+// after: User-generated code
+import 'package:json_annotation/json_annotation.dart';
+
+part 'example.g.dart';
+
+@JsonSerializable(nullable: false)
+class Person {
+  final String firstName;
+  final String lastName;
+  final DateTime dateOfBirth;
+  Person({this.firstName, this.lastName, this.dateOfBirth});
+
+  // No special wiring needed!
+}
+```
+
+```dart
+// after: machine-generated
+part of 'example.dart';
+
+class Person {
+  // Note: instead of a factory constructor, a "normal" constructor could be
+  // generated that simply populates the fields!
+  factory Person.fromJson(Map<String, dynamic> json) =>
+    Person(
+      firstName: json['firstName'] as String,
+      lastName: json['lastName'] as String,
+      dateOfBirth: DateTime.parse(json['dateOfBirth'] as String),
+    );
+
+  Map<String, dynamic> toJson() => <String, dynamic>{
+      'firstName': firstName,
+      'lastName': lastName,
+      'dateOfBirth': dateOfBirth.toIso8601String(),
+    };
+}
+```
+
+## Design ideas
+
+### Legal to define the same type multiple times in the same library
+
+The following would be legal:
+
+```dart
+class Person{}
+class Person{}
+class Person{}
+```
+
+...and would be treated as
+
+```dart
+class Person{}
+```
+
+> Why not restrict it to one-definition-per-file? It's common for code
+generators to create one file with the output from several separate generators.
+
+### Class hierarchy and members are merged across partial classes
```

Comment:
What order are mixins applied in? This at least needs to be defined :)

Reply:
The question

```dart
class Person with A{}
class Person with B{}
class Person with C{}
```

Should be treated as 

```dart
class Person with A, B, C{}
```

But if the `Person` definitions are across many parts, whe application of each mixin should be deterministic. Trivial thought: in order they are defined in each file, starting w/ defining library, then each part file in the order in which the directive appears in the deefining library

Reply:
That's consistent, even if somewhat arbitrary. It will be the only place where the order of declarations matter, so I'd prefer to not go there, and simply not allow a partial class part to add a mixin any more than I'd allow it to change the superclass otherwise.